modFastVM/src/org/aion/fastvm/ExecutionContext.java

```diff
@@ -0,0 +1,316 @@
+    public Address getOriginAddress() {
+        return origin;
+    }
+
+    /** @return the transaction caller. */
+    @Override
+    public Address getSenderAddress() {
+        return sender;
+    }
+
+    /** @return the nrg price in current environment. */
+    @Override
+    public long getTransactionEnergyPrice() {
+        if (this.nrgPrice instanceof DataWord) {
+            return ((DataWord) this.nrgPrice).longValue();
+        } else {
+            return ((DoubleDataWord) this.nrgPrice).longValue();
+        }
+    }
+
+    /** @return the nrg limit in current environment. */
+    public long getTransactionEnergy() {
+        return nrg;
+    }
+
+    /** @return the deposited value by instruction/transaction. */
+    @Override
+    public BigInteger getTransferValue() {
+        return callValue.value();
+    }
+
+    /** @return the call data. */
+    @Override
+    public byte[] getTransactionData() {
+        return callData;
+    }
+
+    /** @return the execution stack depth. */
+    @Override
+    public int getTransactionStackDepth() {
+        return depth;
+    }
+
+    /** @return the transaction kind. */
+    @Override
+    public int getTransactionKind() {
+        return kind;
+    }
+
+    /** @return the transaction flags. */
+    @Override
+    public int getFlags() {
+        return flags;
+    }
+
+    /** @return the block's beneficiary. */
+    @Override
+    public Address getMinerAddress() {
+        return blockCoinbase;
+    }
+
+    /** @return the block number. */
+    @Override
+    public long getBlockNumber() {
+        return blockNumber;
+    }
+
+    /** @return the block timestamp. */
+    @Override
+    public long getBlockTimestamp() {
+        return blockTimestamp;
+    }
+
+    /** @return the block energy limit. */
+    @Override
+    public long getBlockEnergyLimit() {
+        return blockNrgLimit;
+    }
+
+    /** @return the block difficulty. */
+    @Override
+    public long getBlockDifficulty() {
+        if (blockDifficulty instanceof DataWord) {
+            return ((DataWord) blockDifficulty).longValue();
+        } else {
+            return ((DoubleDataWord) blockDifficulty).longValue();
+        }
+    }
+
+    /** @return the transaction helper. */
+    @Override
+    public TransactionSideEffects getSideEffects() {
+        return sideEffects;
+    }
+
+    /**
+     * Sets the transaction hash to txHash.
+     *
+     * @param txHash The new transaction hash.
+     */
+    @Override
+    public void setTransactionHash(byte[] txHash) {
+        this.txHash = txHash;
+    }
+
+    /**
+     * Returns the length of the big-endian binary encoding of this ExecutionContext.
+     *
+     * @return the legtn of this ExecutionContext's binary encoding.
+     */
+    private int getEncodingLength() {
+        return ENCODE_BASE_LEN + callData.length;
+    }
+
+    /** @return the original transaction hash. */
+    @Override
+    public byte[] getHashOfOriginTransaction() {
```

Comment:
Origin has just been the word we've used, and it's used in many places. I had originally wanted to change it to original but I'm not sure how much more descriptive value it will add and we also need to make sure we catch every usage so that there's no odd misalignment that gets to be confusing. Probably a separate task, I'll note it during the next re-working of the API